wp-includes/class-wp-service-worker-scripts.php

```diff
@@ -0,0 +1,245 @@
+<?php
+/**
+ * Dependencies API: WP_Service_Worker_Scripts class
+ *
+ * @since 0.2
+ * @package PWA
+ */
+
+/**
+ * Class used to register service workers.
+ *
+ * @since 0.1
+ *
+ * @see WP_Dependencies
+ *
+ * @method WP_Service_Worker_Precaching_Routes_Component precaching_routes()
+ * @method WP_Service_Worker_Caching_Routes_Component    caching_routes()
+ */
+class WP_Service_Worker_Scripts extends WP_Scripts implements WP_Service_Worker_Registry {
+
+	/**
+	 * Service worker components.
+	 *
+	 * @since 0.2
+	 * @var array
+	 */
+	protected $components = array();
+
+	/**
+	 * Service worker component registries.
+	 *
+	 * @since 0.2
+	 * @var array
+	 */
+	protected $registries = array();
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 0.2
+	 *
+	 * @param array $components Optional. Service worker components as $slug => $instance pairs.
+	 *                          Each component must implement `WP_Service_Worker_Component`.
+	 *                          Default empty array.
+	 */
+	public function __construct( $components = array() ) {
+		$this->components = $components;
+		foreach ( $this->components as $slug => $instance ) {
+			if ( $instance instanceof WP_Service_Worker_Registry_Aware ) {
+				$this->registries[ $slug ] = $instance->get_registry();
+			}
+		}
+
+		parent::__construct();
+	}
+
+	/**
+	 * Initialize the class.
+	 */
+	public function init() {
+		foreach ( $this->components as $component ) {
+			$callback = array( $component, 'serve' );
+			$priority = $component->get_priority();
+
+			add_action( 'wp_front_service_worker', $callback, $priority, 1 );
+			add_action( 'wp_admin_service_worker', $callback, $priority, 1 );
+		}
+
+		/**
+		 * Fires when the WP_Service_Worker_Scripts instance is initialized.
+		 *
+		 * @since 0.1
+		 *
+		 * @param WP_Service_Worker_Scripts $scripts Instance to register service worker behavior with.
+		 */
+		do_action_ref_array( 'wp_default_service_workers', array( &$this ) );
+	}
+
+	/**
+	 * Magic call method. Allows accessing component registries.
+	 *
+	 * @since 0.2
+	 *
+	 * @param string $method Method name. Should be the identifier for a component registry.
+	 * @param array  $args   Method arguments.
+	 * @return WP_Service_Worker_Registry|null Registry instance if valid method name, otherwise null.
+	 */
+	public function __call( $method, $args ) {
+		if ( isset( $this->registries[ $method ] ) ) {
+			return $this->registries[ $method ];
+		}
+
+		return null;
+	}
+
+	/**
+	 * Registers a service worker script.
+	 *
+	 * @since 0.2
+	 *
+	 * @param string $handle Handle of the script.
+	 * @param array  $args   {
+	 *     Additional script arguments.
+	 *
+	 *     @type string|callable $src  Required. URL to the source in the WordPress install, or a callback that
+	 *                                 returns the JS to include in the service worker.
+	 *     @type array           $deps An array of registered item handles this item depends on. Default empty array.
```

Comment:
The type here would be updated to `array|string|callable`, no?

Reply:
The type for the parameter? Indeed, that would need to be adjusted, also for `WP_Service_Worker_Precaching_Routes::register()` and `WP_Service_Worker_Caching_Routes::register()` (for those it would be `array|string`). I did implement the logic, but the doc-block still has only `array` there at the moment.